psydac/fem/tensor.py

```diff
@@ -611,31 +620,11 @@ def reduce_degree(self, axes, basis='B'):
         periods = [V.periodic for V in spaces]
         # create new Tensor Vector
         if v.cart:
-
-            tensor_vec = TensorFemSpace(*spaces, comm=v.cart.comm)
             red_cart = v.cart.remove_last_element(axes)
-            v = StencilVectorSpace(red_cart)
-            tensor_vec._vector_space = v
+            tensor_vec = TensorFemSpace(*spaces, cart=red_cart, quad_order=self._quad_order)
         else:
-            tensor_vec = TensorFemSpace(*spaces)
-            v = self._vector_space
-            tensor_vec._vector_space = StencilVectorSpace(npts, v.pads, v.periods)
-            v = tensor_vec._vector_space
-
-        tensor_vec._spaces = tuple(spaces)
-
-       # Compute extended 1D quadrature grids (local to process) along each direction
-       
-        tensor_vec._quad_grids = tuple( FemAssemblyGrid( V,s,e,quad_order=q, pad=q )
-                                  for V,s,e,q in zip( spaces, v.starts, v.ends, self.degree ) )
-
-        tensor_vec._element_starts, tensor_vec._element_ends =  self._element_starts, self._element_ends
-
-        # Compute limits of eta_0, eta_1, eta_2, etc... in subdomain local to process
-        tensor_vec._eta_limits = tuple( (space.breaks[s], space.breaks[e+1])
-        for s,e,space in zip( self._element_starts, self._element_ends, spaces ) )
```

Comment:
Is this code obsolete?

Reply:
I think it might be by now (the code is very similar to the constructor, and hence I thought it would be a good idea to unify as much as possible there; so that future changes are easier doable):

* by initializing the new `TensorFemSpace` with the new `cart`, we only really change the `self.global_element_starts`, `self.global_element_ends` parameters (which are currently not used anywhere, even not in some branches I checked).
* the `_quad_grids` assembly should be the same as when we initialize with the new cart anyways here (we have pads now, after all)
* The `_element_starts` and `_element_ends` will still do their job as described (no double computation), as long as the `FemAssemblyGrid` works by distributing each index (from the `indices` array) onto a unique process (they may change though by reducing the degree). Cf. also the test included in #132 which checks exactly this. Besides, these parameters are currently only used for plotting, and computing the `_eta_limits`, as it seems.
* The `_eta_limits` then are constructed in the same way as in the constructor, i.e. if the previous statements are true, then the `_eta_limits` should be correct as well. Also, it is currently not used outside the `TensorFemSpace` class, as far as I checked.

Of course, if I may have forgotten something (and some of these changes were incorrect), I can revert some of these changes; like the `_element_starts` and `_element_ends`, if necessary.

Reply:
Thanks, I've just double checked and I think that you are right.